eknrenderer/eknr-renderer.c

```diff
@@ -0,0 +1,1306 @@
+                               renderer_mustache_data_free)
+
+static void
+free_mustache_template (mustache_template_t *template)
+{
+  mustache_api_t api = {
+    .freedata = NULL
+  };
+
+  mustache_free (&api, template);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (mustache_template_t, free_mustache_template)
+
+static uintmax_t
+_renderer_read_from_closure (mustache_api_t *api,
+                             void           *userdata,
+                             char           *buffer,
+                             uintmax_t       buffer_size)
+{
+  RendererMustacheData *data = userdata;
+  return mustache_std_strread (api, (void *) &data->input, buffer, buffer_size);
+}
+
+static uintmax_t
+_renderer_write_to_closure (mustache_api_t *api,
+                            void           *userdata,
+                            const char     *buffer,
+                            uintmax_t       buffer_size)
+{
+  RendererMustacheData *data = userdata;
+  return mustache_std_strwrite (api, (void *) &data->output, (char *) buffer, buffer_size);
+}
+
+static const char *
+_lookup_in_gvariant_dict (RendererMustacheData *data,
+                          const char           *text,
+                          mustache_api_t       *api)
+{
+  g_autoptr(GVariant) value_v = g_variant_dict_lookup_value (data->variables, text, G_VARIANT_TYPE_STRING);
+
+  if (value_v == NULL)
+    {
+      g_autofree char *msg = g_strdup_printf ("No such variable %s", text);
+      (*api->error) (api, data, __LINE__, msg);
+      return NULL;
+    }
+
+  return g_variant_get_string (value_v, NULL);
+}
+
+static const char *
+html_entity_for_char (char matched_ent)
+{
+  switch (matched_ent) {
+    case '&':
+      return "&amp;";
+    case '<':
+      return "&lt;";
+    case '>':
+      return "&gt;";
+    case '"':
+      return "&quot;";
+    case '\'':
+      return "&#39;";
+    case '/':
+      return "&#x2F;";
+    default:
+      return NULL;
+  }
+}
+
+static gboolean
+escape_value_eval_callback (const GMatchInfo       *match_info,
+                            GString                *result,
+                            G_GNUC_UNUSED gpointer  user_data)
+{
+  g_autofree char *matched = g_match_info_fetch (match_info, 0);
+  char matched_ent = matched[0];
+  const char *html_entity = html_entity_for_char (matched_ent);
+
+  if (html_entity == NULL)
+    return TRUE;
+
+  g_string_append (result, html_entity);
+
+  /* From the documentation, returning FALSE continues the replacement
+   * process, returning TRUE stops it. It isn't entirely clear, but it seems
+   * that prematurely stopping the replacement process is an error case. */
+  return FALSE;
+}
+
+/**
+ * eknr_escape_html:
+ * @html: The string to escape
+ * @error: A #GError
+ *
+ * Escape HTML, replacing characters with HTML entities.
+ *
+ * Returns: (transfer full): The escaped HTML document, or %NULL on error.
+ */
+char *
+eknr_escape_html (const char *html,
```

Comment:
I believe what this function does should already be covered by `g_markup_escape_text()`.

Reply:
Thanks!